sdk/openai/openai/MIGRATION.md

```diff
@@ -0,0 +1,504 @@
+
+### Microsoft Entra ID (Formerly known as Azure Active Directory)
+
+There are several ways to authenticate with the Azure OpenAI service using Microsoft Entra ID tokens. The default way is to use the `DefaultAzureCredential` class from the `@azure/identity` package but your application may be using a different credential class. For the purposes of this guide, we will assume that you are using the `DefaultAzureCredential` class. A credential can be created as follows:
+
+```typescript
+import { DefaultAzureCredential } from "@azure/identity";
+const credential = new DefaultAzureCredential();
+```
+
+Which is then passed to the second argument of the client constructors of `OpenAIClient` and `AssistantsClient`.
+
+However, in order to authenticate the `AzureOpenAI` client, we need to use the `getBearerTokenProvider` function from the `@azure/identity` package. This function creates a token provider that `AzureOpenAI` uses internally to obtain tokens for each request. The token provider is created as follows:
+
+```typescript
+import { DefaultAzureCredential, getBearerTokenProvider } from "@azure/identity";
+const credential = new DefaultAzureCredential();
+const scope = "https://cognitiveservices.azure.com/.default";
+const azureADTokenProvider = getBearerTokenProvider(credential, scope);
+```
+
+`azureADTokenProvider` will be passed to the options object when creating the `AzureOpenAI` client.
+
+### (⚠️ Highly Discouraged) API Key
+
+API keys are not recommended for production use because they are less secure than other authentication methods. However, if you are using an API key to authenticate `OpenAIClient` or `AssistantsClient`, an `AzureKeyCredential` object must have been created as follows:
+
+```typescript
+import { AzureKeyCredential } from "@azure/openai";
+const apiKey = new AzureKeyCredential("your API key");
+```
+
+Authenticating `AzureOpenAI` with an API key is as simple as setting the `AZURE_OPENAI_API_KEY` environment variable or as setting the `apiKey` string property in the options object when creating the `AzureOpenAI` client.
+
+## Constructing the client
+
+`OpenAIClient` and `AssistantsClient` are constructed similarly as follows:
+
+Original code:
+```typescript
+import { OpenAIClient } from "@azure/openai";
+const endpoint = "Your Azure OpenAI resource endpoint";
+const client = new OpenAIClient(endpoint, credential);
+```
+
+If not set, the API version defaults to the last known one before the release of the client. Also, the client is not locked to a single model deployment, meaning that the deployment name has to be passed to each method that requires it.
+
+On the other hand, the `AzureOpenAI` client is constructed as follows:
+
+Migrated code:
+```typescript
+import { AzureOpenAI } from "openai";
```

Comment:
Yes it doesn't because we recommend to keep it in an environment variable. `AzureOpenAI` will automatically load it from there. See the paragraph right below this code snippet. If the customer wants to hardcode it, they can set the `endpoint` option but that is not recommended.